riscv/src/small_field/runtime.rs

```diff
@@ -0,0 +1,354 @@
+            alias: alias.map(|s| s.to_string()),
+            instance_name: instance_name.to_string(),
+            arguments: arguments.into_iter().map(|s| s.to_string()).collect(),
+            instructions: instructions
+                .into_iter()
+                .map(|s| parse_instruction_declaration(s.as_ref()))
+                .collect(),
+            extra_registers,
+            init_call: init_call
+                .into_iter()
+                .map(|s| parse_function_statement(s.as_ref()))
+                .collect(),
+        };
+        assert!(
+            self.submachines
+                .insert(instance_name.to_string(), subm)
+                .is_none(),
+            "submachine {instance_name} already present"
+        );
+    }
+
+    fn add_syscall<S: AsRef<str>, I: IntoIterator<Item = S>>(
+        &mut self,
+        syscall: Syscall,
+        implementation: I,
+    ) {
+        let implementation = SyscallImpl(
+            implementation
+                .into_iter()
+                .map(|s| parse_function_statement(s.as_ref()))
+                .collect(),
+        );
+
+        if self.syscalls.insert(syscall, implementation).is_some() {
+            panic!("duplicate syscall {syscall}");
+        }
+    }
+
+    fn with_poseidon(self, _continuations: bool) -> Self {
+        todo!()
+    }
+
+    pub fn has_submachine(&self, name: &str) -> bool {
+        self.submachines.contains_key(name)
+    }
+
+    fn with_arith(self) -> Self {
+        todo!()
+    }
+
+    pub fn submachines_init(&self) -> Vec<String> {
+        self.submachines
+            .values()
+            .flat_map(|m| m.init_call.iter())
+            .map(|s| s.to_string())
+            .collect()
+    }
+
+    pub fn submachines_import(&self) -> String {
+        self.submachines.values().map(|m| m.import()).join("\n")
+    }
+
+    pub fn submachines_declare(&self) -> String {
+        self.submachines
+            .values()
+            .map(|m| m.declaration())
+            .join("\n")
+    }
+
+    pub fn submachines_instructions(&self) -> Vec<String> {
+        self.submachines
+            .values()
+            .flat_map(|m| m.instructions.iter())
+            .map(|s| s.to_string())
+            .collect()
+    }
+
+    pub fn submachines_extra_registers(&self) -> Vec<String> {
+        let count = self
+            .submachines
+            .values()
+            .map(|m| m.extra_registers)
+            .max()
+            .unwrap_or(0);
+
+        (0..count)
+            .map(|i| format!("reg {EXTRA_REG_PREFIX}{i};"))
+            .collect()
+    }
+
+    pub fn ecall_handler(&self) -> Vec<String> {
+        let ecall = [
+            "// ecall handler".to_string(),
+            "__ecall_handler:".to_string(),
+        ]
+        .into_iter();
+
+        let jump_table = self.syscalls.keys().map(|s| {
+            let s32_h = ((*s as u32) >> 16) as u16;
+            let s32_l = ((*s as u32) & 0xffff) as u16;
```

Comment:
there are utils for this `u32_low` `u32_high`

Reply:
done